### DISCUSSION TITLE
A Step Closer to Understanding "$dynamicRef"

### DISCUSSION THREAD

**Discussion by watuwo** [json-schema-org/community#281] [Q&A]
I was trying to understand the dereferencing of `"$dynamicRef"` directly from the [specification](https://json-schema.org/draft/2020-12/json-schema-core.html#section-8.2.3.2-4). I failed horribly.

Luckily I have discovered [this question](https://github.com/json-schema-org/community/discussions/264) which mentioned the [slides here](https://github.com/orgs/json-schema-org/discussions/220).

The slides are great! Now I believe I understand how it should work. Now I also believe that understanding this from the specification was completely hopeless (at least for me). The pros at ajv also seem confused (people have created issues, e.g. [issue 1573](https://github.com/ajv-validator/ajv/issues/1573), [issue 1964](https://github.com/ajv-validator/ajv/issues/1964); the current ajv version says [this](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L115) is valid).

Stuff that confused *me*:

- The core specification tells us to replace the initial URI with "the URI [...] for the outermost schema resource in the dynamic scope [...]". In [this test case](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L85) from the test suite: is any URI of the [intended target schema](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L90) ever "the URI [...] for the outermost schema resource in the dynamic scope [...]"? Now I see that it is supposed to work differently: We have to search for a `"$dynamicAnchor"` in each upper scope and replace the URI with a `staticBaseUri + "#" + dynamicAnchor` URI. Also, as seen in the test case, the target does not have to be a "resource" (in the sense of `"$id"`-owner or root).
- The specification tells us to replace URIs but it does not tell us what to do with the replacement URI (presumably treat analogously to `"$ref"`).
- If I add an `"$id"` to the intended target schema in the previously mentioned test case (add `"$id"` [here](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L91)), then [@hyperjump/json-schema](https://www.npmjs.com/package/@hyperjump/json-schema) suddenly finds that the instance [["foo", 42]](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L115) is also valid. The slides address this too. We have to stop the search for `"$dynamicAnchor"` at schemas owning an `"$id"`.
- Words like "current" or "created" refer to contexts that have not explicitly been established. This is ok, but when I am already confused, I start to doubt everything.
- The example in the [appendix C of the specification](https://json-schema.org/draft/2020-12/json-schema-core.html#recursive-example) is a "good weather example" because it contains a `"$dynamicAnchor"` at the roots and doesn't contain inner `"$id"`s. The example is a good introduction but *I* couldn't extract from it how the dereferencing is supposed to work in general.

I am not going to request a specification update or any other action (other people might understand it; I know it's work). Maybe the slides could be mentioned in a `cref` in the appendix for now? Maybe I should just be patient :).

Thank you for the slides, the forum, the above mentioned forum question and its answer!

I do not expect an answer. If I have chosen the wrong channel for this kind of contribution, please tell me and I will try to delete it.

**Comment by jdesrosiers** [comment]
Thanks for the feedback. We're aware that this is a confusing topic, but it's helpful to hear what specifically you found confusing. We were working on it, but that's been put on hold while something more important has taken priority. I'm looking forward to getting back to that when I can. Things you can expect in the future are simplifying the behavior of `$dynamicRef`, improving how the spec describes dynamic scoping, and user friendly documentation on the [Understanding JSON Schema](https://json-schema.org/understanding-json-schema/structuring.html#extending-recursive-schemas) site.

In the meantime, people have found the [official test suite](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json) to be a good source of examples that help them understand the feature. You might also find this [StackOverflow answer](https://stackoverflow.com/questions/69728686/explanation-of-dynamicref-dynamicanchor-in-json-schema-as-opposed-to-ref-and/69774520#69774520) to be a useful introduction that doesn't get too bogged down in the technical details.
